README.md
[![banner](https://raw.githubusercontent.com/nevermined-io/assets/main/images/logo/banner_logo.png)](https://nevermined.io)

# Python API for Nevermined Data platform

> Python SDK for connecting with Nevermined Data Platform
> [nevermined.io](https://nevermined.io)

[![PyPI](https://img.shields.io/pypi/v/nevermined-sdk-py.svg)](https://pypi.org/project/nevermined-sdk-py/)
[![Python package](https://github.com/nevermined-io/sdk-py/workflows/Python%20package/badge.svg)](https://github.com/nevermined-io/sdk-py/actions)
---

## Table of Contents

  - [Features](#features)
  - [Prerequisites](#prerequisites)
  - [Quick-start](#quick-start)
     - [Usage:](#usage)
  - [Configuration](#configuration)
  - [Development](#development)
        - [Testing](#testing)
        - [New Version / New Release](#new-version-new-release)
  - [License](#license)

---

## Features

Sdk-py include the methods to make easy the connection with contracts deployed in different networks.
This repository include also the methods to encrypt and decrypt information using the Parity Secret Store.

## Prerequisites

Python 3.6

To use the data transfer proof, a C++ library is required:

```bash
git clone https://github.com/nevermined-io/rapidsnark
cd rapidsnark
git submodule update --init --recursive
sh ./scripts/install-linux.sh
```

## Quick-start

Install sdk:

```
pip install nevermined-sdk-py
```

### Usage:

```python
import os
import time

from nevermined_sdk_py import (
    Nevermined,
    ConfigProvider,
    Config,
    Metadata,
    Account
)

ConfigProvider.set_config(Config('config.ini'))
# Make a new instance of Nevermined
nevermined = Nevermined() # or Nevermined(Config('config.ini'))
config = nevermined.config
# make account instance, assuming the ethereum account and password are set 
# in the config file `config.ini`
account = nevermined.accounts.list()[0]
# or 
account = Account(config.parity_address, config.parity_password)

# PUBLISHER
# Let's start by registering an asset in the Nevermined network
metadata = Metadata.get_example()

# access and service endpoints require `gateway.url` is set in the config file
# or passed to Nevermined instance in the config_dict.
# define the services to include in the new asset DDO

ddo = nevermined.assets.create(metadata, account)

# Now we have an asset registered, we can verify it exists by resolving the did
_ddo = nevermined.assets.resolve(ddo.did)
# ddo and _ddo should be identical

# CONSUMER
# search for assets
asset_ddo = nevermined.assets.search('Nevermined protocol')[0]
# Need some ocean tokens to be able to order assets
nevermined.accounts.request_tokens(account, 10)

# Start the purchase/access request. This will automatically make a payment from the specified account.
consumer_account = nevermined.accounts.list()[1]
service_agreement_id = nevermined.assets.order(asset_ddo.did, 0, consumer_account)

# after a short wait (seconds to minutes) the asset data files should be available in the `downloads.path` defined in config
# wait a bit to let things happen
time.sleep(20)

# Asset files are saved in a folder named after the asset id
dataset_dir = os.path.join(nevermined.config.downloads_path, f'datafile.{asset_ddo.asset_id}.0')
if os.path.exists(dataset_dir):
    print('asset files downloaded: {}'.format(os.listdir(dataset_dir)))

```

## Configuration

```python
config_dict = {
    'keeper-contracts': {
        # Point to an Ethereum RPC client. Note that sdk learns the name of the network to work with from this client.
        'keeper.url': 'http://localhost:8545',
        # Specify the keeper contracts artifacts folder (has the smart contracts definitions json files). When you
        # install the package, the artifacts are automatically picked up from the `keeper-contracts` Python
        # dependency unless you are using a local ethereum network.
        'keeper.path': 'artifacts',
        'secret_store.url': 'http://localhost:12001',
        'parity.url': 'http://localhost:8545',
        'parity.address': '',
        'parity.password': '',

    },
    'resources': {
        # Metadata is the metadata store. It stores the assets DDO/DID-document
        'metadata.url': 'http://localhost:5000',
        # Gateway is the publisher's agent. It serves purchase and requests for both data access and compute services
        'gateway.url': 'http://localhost:8030',
        # points to the local database file used for storing temporary information (for instance, pending service agreements).
        'storage.path': 'sdk.db',
        # Where to store downloaded asset files
        'downloads.path': 'access-downloads'
    }
}

```

In addition to the configuration file, you may use the following environment variables (override the corresponding configuration file values):

- KEEPER_PATH
- KEEPER_URL
- GAS_LIMIT
- METADATA_URL

## Development

1. Set up a virtual environment

    ```bash
    virtualenv venv -p python3.6
    source venv/bin/activate 
    ```

1. Install requirements

    ```
    pip install -r requirements_dev.txt
    ```

1. Create the local testing environment using [nevermined-tools](https://github.com/nevermined-io/tools). Once cloned that repository, you can start the cluster running:

    ```
    ./start_nevermined.sh --latest --no-gateway --no-common --local-spree-node
    ```

    It runs a Nevermined Metadata node and an Ethereum RPC client. For details, read `docker-compose.yml`.

1. Create local configuration file

    ```
    cp config.ini config_local.ini
    ```

   `config_local.ini` is used by unit tests.

1. Copy keeper artifacts

    A bash script is available to copy keeper artifacts into this file directly from a running docker image. This script needs to run in the root of the project.
    The script waits until the keeper contracts are deployed, and then copies the artifacts.

    ```
    ./scripts/wait_for_migration_and_extract_keeper_artifacts.sh
    ```

    The artifacts contain the addresses of all the deployed contracts and their ABI definitions required to interact with them.


#### Testing

Automatic tests are setup via Github actions
Our test use pytest framework.

```bash
export PARITY_ADDRESS=0x00bd138abd70e2f00903268f3db08f2d25677c9e
export PARITY_PASSWORD=node0
export PARITY_KEYFILE=tests/resources/data/key_file_2.json
export PARITY_ADDRESS1=0x068ed00cf0441e4829d9784fcbe7b9e26d4bd8d0
export PARITY_PASSWORD1=secret
export PARITY_KEYFILE1=tests/resources/data/key_file_1.json
```

#### New Version / New Release

See [RELEASE_PROCESS.md](RELEASE_PROCESS.md)

##Attribution
This project is based in the [Ocean Protocol Squid-py](https://github.com/oceanprotocol/squid-py). It keeps the same Apache v2 License and adds some improvements.


## License

```text
Copyright 2020 Keyko GmbH.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v0.14.0](https://github.com/nevermined-io/sdk-py/compare/v0.13.0...v0.14.0)

> 10 June 2022

- update to latest contracts [`#65`](https://github.com/nevermined-io/sdk-py/pull/65)

#### [v0.13.0](https://github.com/nevermined-io/sdk-py/compare/v0.12.0...v0.13.0)

> 6 June 2022

- Add support to marketplace api and contracts 2.0 [`#64`](https://github.com/nevermined-io/sdk-py/pull/64)
- contracts dependency [`#63`](https://github.com/nevermined-io/sdk-py/pull/63)
- Added support for marketplace api [`59755d6`](https://github.com/nevermined-io/sdk-py/commit/59755d6a2c2a9e1e6589f31ba8609ea24224d4a5)
- Several more fixes to conditions and tests [`7366046`](https://github.com/nevermined-io/sdk-py/commit/736604655b360b054356733164538cd554f8dd30)
- inital support for contracts 2.0 [`f26507c`](https://github.com/nevermined-io/sdk-py/commit/f26507c34777e808e0dba960dccd74bc9c824e61)

#### [v0.12.0](https://github.com/nevermined-io/sdk-py/compare/v0.11.2...v0.12.0)

> 9 February 2022

- Upgrade to contracts 1.3.6 [`#62`](https://github.com/nevermined-io/sdk-py/pull/62)
- Adding v0.11.2 Changelog updates [`538e3d2`](https://github.com/nevermined-io/sdk-py/commit/538e3d23c61385ebd00462f659249dbf5d0a8a6a)
- fix typo [`4bafbf5`](https://github.com/nevermined-io/sdk-py/commit/4bafbf53b84f24bf09e385e3310fd0281c88324e)
- add missing test dependency [`8bd744a`](https://github.com/nevermined-io/sdk-py/commit/8bd744a5e073e22c7578533310d338b8dfcb82c2)

#### [v0.11.2](https://github.com/nevermined-io/sdk-py/compare/v0.11.1...v0.11.2)

> 1 February 2022

- contract lib version [`#61`](https://github.com/nevermined-io/sdk-py/pull/61)
- Adding v0.11.1 Changelog updates [`74a4ca0`](https://github.com/nevermined-io/sdk-py/commit/74a4ca08b6528ee92c4bf361e884b88e738321d4)
- fix test [`6813c96`](https://github.com/nevermined-io/sdk-py/commit/6813c96d7bf4d249739c2f873e25f9d239ad630c)
- test response [`99bb9a1`](https://github.com/nevermined-io/sdk-py/commit/99bb9a142704f8d38de55513bfec4f7a39549c17)

#### [v0.11.1](https://github.com/nevermined-io/sdk-py/compare/v0.11.0...v0.11.1)

> 24 January 2022

- commons utils version [`#60`](https://github.com/nevermined-io/sdk-py/pull/60)
- Adding v0.11.0 Changelog updates [`1e88726`](https://github.com/nevermined-io/sdk-py/commit/1e8872688c7181d502d2647f81cfc9786f771f22)

#### [v0.11.0](https://github.com/nevermined-io/sdk-py/compare/v0.10.7...v0.11.0)

> 22 December 2021

- Updated version of contracts-lib [`#59`](https://github.com/nevermined-io/sdk-py/pull/59)
- Adding data transfer proofs here [`#50`](https://github.com/nevermined-io/sdk-py/pull/50)
- key transfer working here too [`23475ed`](https://github.com/nevermined-io/sdk-py/commit/23475ed511855033c0f6ef3afa6d739efd797035)
- timeout params [`93ec664`](https://github.com/nevermined-io/sdk-py/commit/93ec664fdbd80d5acd96852ac5664a8686a3868e)
- Added NFTUpgradeable changes [`4475013`](https://github.com/nevermined-io/sdk-py/commit/4475013e30c9e021b754edd3ae9c5fc9f3edfdd1)

#### [v0.10.7](https://github.com/nevermined-io/sdk-py/compare/v0.10.6...v0.10.7)

> 8 October 2021

- Make AccessProof contracts optional [`#55`](https://github.com/nevermined-io/sdk-py/pull/55)
- Adding v0.10.6 Changelog updates [`be0f22f`](https://github.com/nevermined-io/sdk-py/commit/be0f22f0a0412609aa77cbbd45bb6bd35dd3dc0a)

#### [v0.10.6](https://github.com/nevermined-io/sdk-py/compare/v0.10.5...v0.10.6)

> 8 October 2021

- Set tokenAddress when creating an asset [`#54`](https://github.com/nevermined-io/sdk-py/pull/54)
- Fix failing test [`f7f100c`](https://github.com/nevermined-io/sdk-py/commit/f7f100c69d5cf2b10f1527f5face828cec91798c)
- Set default tokenAddress [`0717a00`](https://github.com/nevermined-io/sdk-py/commit/0717a00716b01281bdce83a7c82e525bccfd0b35)
- bump version [`9257801`](https://github.com/nevermined-io/sdk-py/commit/9257801a0c429231748602e43350f318dcff94b6)

#### [v0.10.5](https://github.com/nevermined-io/sdk-py/compare/v0.10.4...v0.10.5)

> 30 September 2021

- Feature/contracts 1.1.8 [`#52`](https://github.com/nevermined-io/sdk-py/pull/52)
- support aurora networks [`#51`](https://github.com/nevermined-io/sdk-py/pull/51)
- Bump version: 0.10.4 → 0.10.5 [`900b2cf`](https://github.com/nevermined-io/sdk-py/commit/900b2cf8cd7f33ed3daead72aea6d8193de69a27)
- Upgrade version to 1.1.8 [`a871249`](https://github.com/nevermined-io/sdk-py/commit/a871249e2000016f5fa9b7e1e8b935208f409221)

#### [v0.10.4](https://github.com/nevermined-io/sdk-py/compare/v0.10.3...v0.10.4)

> 22 September 2021

- Bump version: 0.10.3 → 0.10.4 [`083f243`](https://github.com/nevermined-io/sdk-py/commit/083f2439390fb773c21585bd630cc5b4702988ea)
- update [`fde6e4c`](https://github.com/nevermined-io/sdk-py/commit/fde6e4c0c012cbfb32a18577fd2c62d5d2b6ad92)

#### [v0.10.3](https://github.com/nevermined-io/sdk-py/compare/v0.10.2...v0.10.3)

> 21 September 2021

- Adding v0.10.2 Changelog updates [`4ef3089`](https://github.com/nevermined-io/sdk-py/commit/4ef3089dddda4123497658d9e104d5de17424618)
- Bump version: 0.10.2 → 0.10.3 [`07d7778`](https://github.com/nevermined-io/sdk-py/commit/07d7778fb74224167d5eb39c59e59ab7ff6c434a)
- fix version [`5e4fcb5`](https://github.com/nevermined-io/sdk-py/commit/5e4fcb57d633776caaaafabfd81eadd109c1d9b5)

#### [v0.10.2](https://github.com/nevermined-io/sdk-py/compare/v0.10.1...v0.10.2)

> 9 September 2021

- Test E2E with mumbai [`#49`](https://github.com/nevermined-io/sdk-py/pull/49)
- Cleaned up test config [`6b3ea86`](https://github.com/nevermined-io/sdk-py/commit/6b3ea8639e1cc4423a83ccac394004a2d67922c5)
- Adding v0.10.1 Changelog updates [`5f009bf`](https://github.com/nevermined-io/sdk-py/commit/5f009bf32e077acb7a269000949fad72e1d5b7da)

nevermind_sdk_py/config.py
import configparser
import logging
import os
import site
from pathlib import Path

DEFAULT_KEEPER_HOST = 'localhost'
DEFAULT_KEEPER_PORT = 8545
DEFAULT_KEEPER_URL = 'http://localhost:8545'
DEFAULT_KEEPER_PATH = 'artifacts'
DEFAULT_GAS_LIMIT = 4000000
DEFAULT_NAME_METADATA_URL = 'http://localhost:5000'
DEFAULT_STORAGE_PATH = 'squid_py.db'

NAME_KEEPER_URL = 'keeper.url'
NAME_KEEPER_PATH = 'keeper.path'
NAME_GAS_LIMIT = 'gas_limit'
NAME_METADATA_URL = 'metadata.url'
NAME_STORAGE_PATH = 'storage.path'
NAME_AUTH_TOKEN_MESSAGE = 'auth_token_message'
NAME_AUTH_TOKEN_EXPIRATION = 'auth_token_expiration'

NAME_SECRET_STORE_URL = 'secret_store.url'
NAME_PARITY_URL = 'parity.url'
NAME_PARITY_ADDRESS = 'parity.address'
NAME_PARITY_PASSWORD = 'parity.password'

environ_names = {
    NAME_KEEPER_URL: ['KEEPER_URL', 'Keeper URL'],
    NAME_KEEPER_PATH: ['KEEPER_PATH', 'Path to the keeper contracts'],
    NAME_GAS_LIMIT: ['GAS_LIMIT', 'Gas limit'],
    NAME_METADATA_URL: ['METADATA_URL', 'Metadata URL'],
    NAME_STORAGE_PATH: ['STORAGE_PATH', 'Path to the local database file'],
    NAME_AUTH_TOKEN_MESSAGE: ['AUTH_TOKEN_MESSAGE',
                              'Message to use for generating user auth token'],
    NAME_AUTH_TOKEN_EXPIRATION: ['AUTH_TOKEN_EXPIRATION',
                                 'Auth token expiration time expressed in seconds'],
    NAME_SECRET_STORE_URL: ['SECRET_STORE_URL', 'Secret Store URL'],
    NAME_PARITY_URL: ['PARITY_URL', 'Parity URL'],
    NAME_PARITY_ADDRESS: ['PARITY_ADDRESS', 'Parity address'],
    NAME_PARITY_PASSWORD: ['PARITY_PASSWORD', 'Parity password'],
}

config_defaults = {
    'keeper-contracts': {
        NAME_KEEPER_URL: DEFAULT_KEEPER_URL,
        NAME_KEEPER_PATH: DEFAULT_KEEPER_PATH,
        NAME_GAS_LIMIT: DEFAULT_GAS_LIMIT,
        NAME_SECRET_STORE_URL: '',
        NAME_PARITY_URL: '',
        NAME_PARITY_ADDRESS: '',
        NAME_PARITY_PASSWORD: '',
    },
    'resources': {
        NAME_METADATA_URL: DEFAULT_NAME_METADATA_URL,
        NAME_STORAGE_PATH: DEFAULT_STORAGE_PATH,
        NAME_AUTH_TOKEN_MESSAGE: '',
        NAME_AUTH_TOKEN_EXPIRATION: ''
    }
}


class Config(configparser.ConfigParser):
    """Class to manage the nevermind-sdk-py configuration."""

    def __init__(self, filename=None, options_dict=None, **kwargs):
        """
        Initialize Config class.

        Options available:

        [keeper-contracts]
        keeper.url = http://localhost:8545                            # Keeper-contracts url.
        keeper.path = artifacts                                       # Path of json abis.
        secret_store.url = http://localhost:12001                     # Secret store url.
        parity.url = http://localhost:8545                            # Parity client url.
        parity.address = 0x00bd138abd70e2f00903268f3db08f2d25677c9e   # Partity account address.
        parity.password = node0                                       # Parity account password.
        [resources]
        metadata.url = http://localhost:5000                          # Metadata url.
        gateway.url = http://localhost:8030                           # Gateway url.
        storage.path = squid_py.db                                    # Path of sla back-up storage.

        :param filename: Path of the config file, str.
        :param options_dict: Python dict with the config, dict.
        :param kwargs: Additional args. If you pass text, you have to pass the plain text
        configuration.
        """
        configparser.ConfigParser.__init__(self)

        self.read_dict(config_defaults)
        self._web3_provider = None
        self._section_name = 'keeper-contracts'
        self._logger = logging.getLogger('config')

        if filename:
            self._logger.debug(f'Config: loading config file {filename}')
            with open(filename) as fp:
                text = fp.read()
                self.read_string(text)
        else:
            if 'text' in kwargs:
                self.read_string(kwargs['text'])

        if options_dict:
            self._logger.debug(f'Config: loading from dict {options_dict}')
            self.read_dict(options_dict)

        self._load_environ()

    def _load_environ(self):
        for option_name, environ_item in environ_names.items():
            value = os.environ.get(environ_item[0])
            if value is not None:
                self._logger.debug(f'Config: setting environ {option_name} = {value}')
                self.set(self._section_name, option_name, value)

    @property
    def keeper_path(self):
        """Path where the keeper-contracts artifacts are allocated."""
        keeper_path_string = self.get(self._section_name, NAME_KEEPER_PATH)
        path = Path(keeper_path_string).expanduser().resolve()
        # TODO: Handle the default case and make default empty string
        # assert path.exists(), "Can't find the keeper path: {} ({})"..format(keeper_path_string,
        # path)
        if os.path.exists(path):
            pass
        elif os.getenv('VIRTUAL_ENV'):
            path = os.path.join(os.getenv('VIRTUAL_ENV'), 'artifacts')
        else:
            path = os.path.join(site.PREFIXES[0], 'artifacts')
        return path

    @property
    def storage_path(self):
        """Path to save the current execution of the service agreements and restart if needed."""
        return self.get('resources', NAME_STORAGE_PATH)

    @property
    def keeper_url(self):
        """URL of the keeper. (e.g.): http://mykeeper:8545."""
        return self.get(self._section_name, NAME_KEEPER_URL)

    @property
    def gas_limit(self):
        """Ethereum gas limit."""
        return int(self.get(self._section_name, NAME_GAS_LIMIT))

    @property
    def metadata_url(self):
        """URL of metadata component. (e.g.): http://mymetadata:5000."""
        return self.get('resources', NAME_METADATA_URL)

    @property
    def secret_store_url(self):
        """URL of the secret store component. (e.g.): http://mysecretstore:12001."""
        return self.get(self._section_name, NAME_SECRET_STORE_URL)

    @property
    def parity_url(self):
        """URL of parity client. (e.g.): http://myparity:8545."""
        return self.get(self._section_name, NAME_PARITY_URL)

    @property
    def parity_address(self):
        """Parity address. (e.g.): 0x00bd138abd70e2f00903268f3db08f2d25677c9e."""
        return self.get(self._section_name, NAME_PARITY_ADDRESS)

    @property
    def parity_password(self):
        """Parity password for your address. (e.g.): Mypass."""
        return self.get(self._section_name, NAME_PARITY_PASSWORD)

    @property
    def downloads_path(self):
        """Path for the downloads of assets."""
        return self.get('resources', 'downloads.path')

    @property
    def auth_token_message(self):
        return self.get('resources', NAME_AUTH_TOKEN_MESSAGE)

    @property
    def auth_token_expiration(self):
        return self.get('resources', NAME_AUTH_TOKEN_EXPIRATION)

    @property
    def web3_provider(self):
        """Web3 provider"""
        return self._web3_provider

    @web3_provider.setter
    def web3_provider(self, web3_provider):
        self._web3_provider = web3_provider
